addons/mail/static/src/core/common/attachment_upload_service.js

```diff
@@ -0,0 +1,190 @@
+                    ...response,
+                    extension: upload.title.split(".").pop(),
+                    originThread: hooker.composer ? undefined : originThread,
+                });
+                if (hooker.composer) {
+                    const index = hooker.composer.attachments.findIndex(({ id }) => id === tmpId);
+                    if (index >= 0) {
+                        hooker.composer.attachments[index] = attachment;
+                    } else {
+                        hooker.composer.attachments.push(attachment);
+                    }
+                }
+                const def = this.deferredByAttachmentId.get(tmpId);
+                this.unlink(this.store.attachments[tmpId]);
+                if (def) {
+                    def.resolve(attachment);
+                    this.deferredByAttachmentId.delete(tmpId);
+                }
+                hooker.onFileUploaded?.();
+                this.hookersByTmpId.delete(tmpId);
+            }
+        );
+        this.fileUploadService.bus.addEventListener(
+            "FILE_UPLOAD_ERROR",
+            ({ detail: { upload } }) => {
+                const tmpId = parseInt(upload.data.get("temporary_id"));
+                if (!this.uploadingAttachmentIds.has(tmpId)) {
+                    return;
+                }
+                this.abortByAttachmentId.delete(tmpId);
+                this.deferredByAttachmentId.delete(tmpId);
+                this.uploadingAttachmentIds.delete(parseInt(tmpId));
+                this.hookersByTmpId.delete(tmpId);
+            }
+        );
+    }
+
+    async unlink(attachment) {
+        const abort = this.abortByAttachmentId.get(attachment.id);
+        const def = this.deferredByAttachmentId.get(attachment.id);
+        if (abort) {
+            abort();
+            def.resolve();
+        }
+        this.abortByAttachmentId.delete(attachment.id);
+        this.deferredByAttachmentId.delete(attachment.id);
+        await this.attachmentService.delete(attachment);
+    }
+
+    async uploadFile(hooker, file, options) {
+        const tmpId = nextId--;
+        this.hookersByTmpId.set(tmpId, hooker);
+        this.uploadingAttachmentIds.add(tmpId);
+        await this.fileUploadService
+            .upload("/mail/attachment/upload", [file], {
+                buildFormData: (formData) => {
+                    this._makeFormData(formData, file, hooker, tmpId, options);
+                },
+            })
+            .catch((e) => {
+                if (e.name !== "AbortError") {
+                    throw e;
+                }
+            });
+        const uploadDoneDeferred = new Deferred();
+        this.deferredByAttachmentId.set(tmpId, uploadDoneDeferred);
+        return uploadDoneDeferred;
+    }
+
+    _makeFormData(formData, file, hooker, tmpId, options) {
+        formData.append("thread_id", hooker.thread.id);
+        formData.append("tmp_url", URL.createObjectURL(file));
```

Comment:
I think that this url has to be revoked because it will prevent the file from being garbage collected.

Reply:
I added `revokeObjectURL` in `AttachmentService@remove(attachment)`, so at least deleted attachments are cleaned.
There's still a risk of keeping object url in RAM, because we don't cleanly remove attachments by calling `AttachmentService@remove(attachment)` enough times... so in theory and in some cases this can still be an issue. IMO not worth refactoring the code yet